### ISSUE TITLE
Failed to spawn tasks in dynamically loaded cdylibs

### ISSUE THREAD

**Issue by seirdb** [stjepang/smol#88]
I’m working on a plugin-based system where I’m dynamically loading .so depending on some runtime-configuration. In those shared objects I’m running an init function which might spawn an async Task, i.e.:
```
async fn init {
   println!(“{:?}”, std::thread::current());
   Task::local(async {
       // some more code lives here.
    }).detach();
}
```
.

This however fails with the following error:
`thread '<unnamed>' panicked at 'cannot spawn a thread-local task if not inside an executor'`

I think the interesting bit of info here is the `thread ‘<unnamed>’`, the `println!()` in the snippet above simple prints `Thread { id: ThreadId(1), name: Some("main") }`.

My guess is that the dynamically loaded library has it’s own private copy of thread-local-storage variables and thus can’t access the `static EXECUTOR`.

I’ve seen the same issue with Tokio. It does work in async-std however.

**Comment by seirdb** [comment]
As a quick sidenote: the smol library does deliver on it’s promise. I could quickly dig down to where the issue probably arises.

**Comment by seirdb** [comment]
I’ve been thinking on this today, and the only way I see this could be implemented is by keeping a lookup table of all local executors for each `thread id`. This lookup table should be a global static, meaning shared across threads.

Looking at the work-stealing code, the same issue would happen there.

I’m willing to work on a POC, if this is something we’d want.

**Comment by stjepang** [comment]
Can you share more - how would that work exactly?

Similar problems were reported with crossbeam-channel and I've concluded that this simply cannot work correctly so now I'm curious :) https://github.com/crossbeam-rs/crossbeam/issues/424

**Comment by seirdb** [comment]
It was nested deeper than I thought. I was thinking `libpthread` would store (references to) data itself. Since it’s linked as a dynamic library, it would thus circumvent the issue with the “double copies” of the `static` variables. Not so.

A minimal example demonstrating the issue can be found [here](https://github.com/seirdb/tls_issue).

Furthermore, the `pthread_key_create()` call doesn’t return the same value when doing multiple calls from the same thread (as explained in the manpages).

**Comment by seirdb** [comment]
I was also baffled that it did work in `async-std`, but on further investigation, it looks that there it was happily spawning new threads, which then receive their own Runtime.

I’m closing this issue as I also don’t see a way around it.

In order to implement it on my side, I’ll seed the plugin code upon initialization with a callback function that’ll spawn the task in the correct main executable context.
